24.0.0.12-beta staging (#4141)
OpenLiberty/blogs: posts/2024-11-19-24.0.0.12-beta.adoc

open-graph-image-alt: Open Liberty Logo
---
= Enhanced message validation for XML Web Services in 24.0.0.12-beta
David Mueller <https://github.com/dmuelle>
:imagesdir: /
:url-prefix:
:url-about: /

The 24.0.0.12-beta release enhances inbound SOAP message validation in XML Web Services to simplify message debugging and make your web services and clients more resilient.

See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/29770
// Contact/Reviewer: neuwerk
// // // // // // // //
[#xmlws]
== Fine-tuning XML Web Services inbound SOAP message validation

Open Liberty's XML Web Services features now support fine-grained message validation for inbound SOAP messages. This enhancement provides more control over message validation options.
In Open Liberty 24.0.0.12-beta, you can configure message validation using new attributes in the `server.xml` file. These attributes are available for the `webService` and `webServiceClient` elements.

[options="header"]

> |=======================

Reviewer: maybe add a "Default" column to specify default for each attr?

| Attribute | Description
| `enableSchemaValidation` | Enable full validation against the XML schema
| `enableDefaultValidation` | Enable or disable default validation for JAXB
| `ignoreUnexpectedElements` | Use default validation while ignoring `UnmarshallExceptions: Unknown Element` errors
|=======================

By using these attributes, you can tailor message validation to your specific needs and improve the security and reliability of your SOAP-based web services.

These attributes require the one of the following XML Web Services features to be enabled in your `server.xml` file:

* xref:{url-prefix}/docs/latest/reference/feature/xmlWS-4.0.html[Jakarta XML Web Services
4.0] (`xmlWS-4.0`)
* xref:{url-prefix}/docs/latest/reference/feature/xmlWS-3.0.html[Jakarta XML Web Services
4.0] (`xmlWS-3.0`)
* xref:{url-prefix}/docs/latest/reference/feature/jaxws-2.2.html[Java Web Services] (`jaxws-2.2`)

You can apply the configuration to web services (`webService`) or web service clients (`webServiceClient`), either globally, or to an individual client or web service implementation.

> The examples in the following sections show how to enable each option. To disable any of these options, set the corresponding attributes to `false`.

Reviewer: > The examples in the following sections show how to enable each option. To disable any of these options, set the corresponding attributes to `false`.

I feel like this is kind of obvious, but OTOH, are we really saying that some of these are enabled by default in some situations, while others are not? So some might be `true` by default, but others `false`, it depends? Feel like we should have more clarity here

=== XML schema validation

You can set the `enableSchemaValidation=true` attribute to provide more insight into JAXB unmarshalling exceptions and make painful message debugging easier. This option is the highest level of XML validation, which provides faster debugging and the most thorough checks on inbound message contents. But it comes with a tradeoff: higher performance cost.

==== Global XML schema validation

The following example shows how to enable XML schema validation for web services globally for your Open Liberty runtime:

[source,xml]
----
<webService enableSchemaValidation="true" />
----

To enable XML schema validation globally for web service clients, set the same attribute on the `webServiceClient` element:

[source,xml]
----
<webServiceClient enableSchemaValidation="true" />
----

==== Targeted XML schema validation

The following example shows how to enable XML schema validation for a particular web service by using the web service port:

[source,xml]
----
<webService portName="<web service port name>"  enableSchemaValidation="true" />
----

> Here `portName`'s value port name of the Web Service implementation you're configuring.  This name comes from your `@WebService(portName=<web service port name>` annotated class.

Reviewer: ```suggestion
The value of `portName` is the port name of the Web Service implementation you're configuring.  This name comes from your `@WebService(portName=<web service port name>` annotated class.
```

> The `web service port name` is the port name of the individual web service implementation.  This name comes from your `@WebService(portName=<web service port name>` annotated class.

Reviewer 2: ```suggestion
Here `portName`'s value port name of the Web Service implementation you're configuring.  This name comes from your `@WebService(portName=<web service port name>` annotated class.
```

Alternatively, you can check the `<wsdl:port ... name="Web Service Port Name">` line in your WSDL file for the port name.

The following example shows how to enable XML schema validation for a specific client service:

[source,xml]
----
<webServiceClient serviceName="<client service name>"  enableSchemaValidation="true" />
----

> The `Client Service Name` is the name of your XML Web Service client. This name comes from your `@WebServiceClient(serviceName=<client service name>` annotated stub class for managed clients.

Reviewer 2: ```suggestion
Here `serviceName`'s value is the name of the Web Service Client you're are configuring. This name comes from your `@WebServiceClient(serviceName=<client service name>` annotated stub class for managed clients.
```

> Here `serviceName`'s value is the name of the Web Service Client you're are configuring. This name comes from your `@WebServiceClient(serviceName=<client service name>` annotated stub class for managed clients.

Reviewer: ```suggestion
The value of `serviceName` is the name of the Web Service Client you're configuring. This name comes from your `@WebServiceClient(serviceName=<client service name>` annotated stub class for managed clients.
```

For unmanaged clients, you can check the `<wsdl:service name="<client service name>"` line in your WSDL file.

=== Default validation

> You can configure the default level of JAXB validation of inbound SOAP Messages with the `enableDefaultValidation=true` attribute. Depending on whether your application is an XML Web Services client or web service, Open Liberty either enables or disables default validation. Default validation is much more efficient than XML schema validation, so enabling it provides basic message validation with lower overhead. Disabling it lets you ignore various unmarshalling errors for problematic messages.

Reviewer 2: ```suggestion
You can configure the default level of JAXB validation of inbound SOAP Messages with the `enableDefaultValidation=true` attribute. Depending on whether your application uses an XML Web Services client or web service or implementation, Open Liberty either enables or disables default validation. Default validation is much more efficient than XML schema validation, so enabling it provides basic message validation with lower overhead. Disabling it lets you ignore various unmarshalling errors for problematic messages.
```

> You can configure the default level of JAXB validation of inbound SOAP Messages with the `enableDefaultValidation=true` attribute. Depending on whether your application uses an XML Web Services client or web service implementation, Open Liberty either enables or disables default validation. Default validation is much more efficient than XML schema validation, so enabling it provides basic message validation with lower overhead. Disabling it lets you ignore various unmarshalling errors for problematic messages.

Reviewer: > Depending on whether your application uses an XML Web Services client or web service implementation, Open Liberty either enables or disables default validation.

does this mean we enable default validation for clients and disable it for services?

> You can configure the default level of JAXB validation of inbound SOAP Messages with the `enableDefaultValidation=true` attribute. Depending on whether your application uses an XML Web Services client or web service or implementation, Open Liberty either enables or disables default validation. Default validation is much more efficient than XML schema validation, so enabling it provides basic message validation with lower overhead. Disabling it lets you ignore various unmarshalling errors for problematic messages.

Reviewer: ```suggestion
You can configure the default level of JAXB validation of inbound SOAP Messages with the `enableDefaultValidation=true` attribute. Depending on whether your application uses an XML Web Services client or web service implementation, Open Liberty either enables or disables default validation. Default validation is much more efficient than XML schema validation, so enabling it provides basic message validation with lower overhead. Disabling it lets you ignore various unmarshalling errors for problematic messages.
```

[source,xml]
----
<webServiceClient serviceName="<client service name>"  enableSchemaValidation="true" />
----

The value of `serviceName` is the name of the Web Service Client you're configuring. This name comes from your `@WebServiceClient(serviceName=<client service name>` annotated stub class for managed clients.
For unmanaged clients, you can check the `<wsdl:service name="<client service name>"` line in your WSDL file.

=== Default validation

You can configure the default level of JAXB validation of inbound SOAP Messages with the `enableDefaultValidation=true` attribute. Default validation is much more efficient than XML schema validation, so enabling it provides basic message validation with lower overhead. Disabling it lets you ignore various unmarshalling errors for problematic messages. Default validation is enabled by default for web services, but disabled for web service clients.

==== Global default validation

The following example shows how to enable default validation globally for web services in your Open Liberty runtime:

[source,xml]
----
<webService enableDefaultValidation="true"/>
----

> Default validation is enabled by default for web service clients. To disable default validation globally for web service clients, set the `enableDefaultValidation="true"` attribute on the `webServiceClient` element.

Reviewer 2: should be `enableDefaultValidation="false"`

Reviewer: oops :(

[source,xml]
----

> <webServiceCLient enableDefaultValidation="false"/>

Reviewer 2: ```suggestion
<webServiceClient enableDefaultValidation="false"/>
```

----

==== Targeted default validation

The following example shows how to enable default validation for a specific web service:

[source,xml]
----
<webService  portName="SayHelloService" enableDefaultValidation="true"/>
----

> Default validation is enabled by default for web service clients. To disable default validation for a specific web service client, set the `enableDefaultValidation="true"` attribute on the `webServiceClient` element and use the `serviceName` attribute to specify the client service.

Reviewer 2: should be enableDefaultValidation="false"
